### PR TITLE
Scheduler algorithm update

### PR DESCRIPTION
**Pull request by nkoskelo**

### CODE REVIEW

File: pytato/distributed/partition.py [hunk 1]
```diff
@@ -522,35 +537,83 @@ def _schedule_comm_batches(
...
+    """
+    Compute a topological sort of the input graph which specifies
+    the tasks that need to be completed before task_i can be scheduled
+    for every i.
+    """
+    locations_visited: Set[CommunicationOpIdentifier] = set()
+    temp_visit: Set[CommunicationOpIdentifier] = set()
+    sorted_list: List[CommunicationOpIdentifier] = []
+
+    def _topo_helper(
```
Comment (inducer) [inline pytato/distributed/partition.py:586]: - Naming.

File: pytato/distributed/partition.py [hunk 2]
```diff
@@ -522,35 +537,83 @@ def _schedule_comm_batches(
...
+# {{{ _topo_sort
+
+
+def _topo_sort(
+        comm_ids_to_needed_comm_ids: CommunicationDepGraph
+        ) -> Tuple[List[CommunicationOpIdentifier], int]:
+    """
+    Compute a topological sort of the input graph which specifies
+    the tasks that need to be completed before task_i can be scheduled
+    for every i.
```
Comment (inducer) [inline pytato/distributed/partition.py:580]: Be more precise about the invariant that's supposed to hold of the output.

**Comment by inducer** [comment]
To cut down on noise in my email, I'm unsubscribing from this PR. When it next needs my attention, please @-mention me or hit the "request review" button. Otherwise, I may not see your messages in a timely manner.

File: pytato/distributed/partition.py [hunk 3]
```diff
@@ -508,49 +508,118 @@ def map_distributed_recv(
...
+    """
+
+    return _schedule_comm_batches(comm_ids_to_needed_comm_ids, cnts)
+
+# }}}
 
 # {{{ _schedule_comm_batches
 
+
 def _schedule_comm_batches(
```
Comment (inducer) [inline pytato/distributed/partition.py:531]: It seems that this produces asymptotically more parts than the prior $O(N^2)$ algorithm. Example:
image

File: pytato/distributed/partition.py [hunk 4]
```diff
@@ -508,49 +508,118 @@ def map_distributed_recv(
...
+        cnts.add(sum([n_visited_in_sort, n_visited_in_scheduling]))
     return comm_batches
 
 # }}}
 
+# {{{ _topo_sort
+
+
+def _topo_sort(
+        comm_ids_to_needed_comm_ids: Mapping[template_t, AbstractSet[template_t]]
```
[the remark this answers is not in the document]
Comment (nkoskelo) [reply pytato/distributed/partition.py:578]: We are removing this function now.

File: pytato/distributed/partition.py (hunk 3 above)
Comment (nkoskelo) [reply pytato/distributed/partition.py]: We now run in $O(N)$ for this example.

File: pytato/distributed/partition.py [hunk 5]
```diff
@@ -509,48 +509,83 @@ def map_distributed_recv(
...
+    """ Calculate the minimum dependendency level needed before a task of
+        type TaskType can be scheduled. We assume that any number of tasks
+        can be scheduled at the same time, and that each task has a constant
+        number of direct dependents.
 
+        The minimum dependency level for a task, i, is defined as
+        1 + the maximum dependency level for its children.
+    """
+    known_vals: Dict[TaskType, int] = {}
+    count: List[int] = [0]
```
[the remark this answers is not in the document]
Comment (nkoskelo) [reply pytato/distributed/partition.py:558]: I had to specify instead that it was a non-local variable.

File: test/test_distributed.py [hunk 6]
```diff
@@ -117,6 +117,81 @@ def _do_test_distributed_execution_basic(ctx_factory):
...
+                needed_ids[j] = {j - branching_factor}
+        _ = _schedule_task_batches(needed_ids, counts)
+        count_list[i] = list(counts)[0]  # python passes by value.
+
+    # Now to do the fitting.
+    coefficients = np.polyfit(sizes, count_list, 4)
+    import numpy.linalg as la
+    nonlinear_norm_frac = la.norm(coefficients[:-2], 2)/la.norm(coefficients, 2)
+    assert nonlinear_norm_frac < 0.0001
+# }}}
```
[the remark this answers is not in the document]
Comment (inducer) [reply test/test_distributed.py:227]: Also test this for expected number of parts.

File: test/test_distributed.py [hunk 7]
```diff
@@ -118,6 +118,115 @@ def _do_test_distributed_execution_basic(ctx_factory):
...
+            else:
+                needed_ids[j] = {j - branching_factor}
+        _, count_list[i] = _schedule_task_batches_counted(needed_ids)
+
+    # Now to do the fitting.
+    coefficients = np.polyfit(sizes, count_list, 4)
+    import numpy.linalg as la
+    nonlinear_norm_frac = la.norm(coefficients[:-2], 2)/la.norm(coefficients, 2)
+    assert nonlinear_norm_frac < 0.0001
+# }}}
```
Comment (inducer) [inline test/test_distributed.py:227]:
```suggestion

# }}}
```

**Comment by inducer** [comment]
Thanks! I've pushed some style fixes, hopefully making the end result easier to read (while mostly leaving the semantics alone). Please take a look.

**Comment by nkoskelo** [comment]
I read through the style changes. They look fine.

**Pull request merged**